server/topic.go

```diff
@@ -2645,41 +2704,27 @@ func (t *Topic) subsCount() int {
 	return len(t.perUser)
 }
 
-func (t *Topic) addSession(s *Session, user types.Uid) types.UidSlice {
-	uids := t.sessions[s]
-	if uids == nil {
-		uids = types.UidSlice{user}
-	} else {
-		uids.Add(user)
+// Add session record. 'user' may be different from s.uid.
+func (t *Topic) addSession(s *Session, asUid types.Uid) bool {
+	if _, ok := t.sessions[s]; ok {
+		return false
 	}
-	t.sessions[s] = uids
-	return uids
+	t.sessions[s] = perSessionData{uid: asUid}
+	return true
 }
 
-// Removes Uid from the session record. If user.IsZero() then removes all UIDs.
-// Returns a slice of removed UIDs (all, just one or nil).
-func (t *Topic) remSession(s *Session, user types.Uid) types.UidSlice {
-	uids := t.sessions[s]
-	if uids == nil {
-		return nil
-	}
-
-	if user.IsZero() {
-		// All uids are removed.
-		delete(t.sessions, s)
-	} else if uids.Rem(user) {
-		// One uid is removed
-		if len(uids) > 0 {
-			t.sessions[s] = uids
-		} else {
-			delete(t.sessions, s)
+// Removes session record if 'asUid' matches subscribed user.
+func (t *Topic) remSession(s *Session, asUid types.Uid) *perSessionData {
```

Comment:
Old implementation assumed multiple uids per session.
Was it wrong (i.e. this PR fixes it up)? If not, in which cases was that possible?

Reply:
It never worked. Session always refused the second subscription to the same topic even if it was on behalf of a different user:
https://github.com/tinode/chat/blob/master/server/session.go#L410

I removed it in https://github.com/tinode/chat/pull/323/commits/e8839894b118ef72e214dadfee320532690320b2

This seems to be a very niche corner case (root user subscribes to the same topic from the same session on behalf of two different users). Supporting it would add too much complexity for such an unimportant feature.

Reply:
Makes sense. Thanks for explaining.